DQMOffline/Trigger/plugins/BPHMonitor.cc

```diff
@@ -0,0 +1,277 @@
+{
+  int nbinsX = binningX.size()-1;
+  std::vector<float> fbinningX(binningX.begin(),binningX.end());
+  float* arrX = &fbinningX[0];
+  int nbinsY = binningY.size()-1;
+  std::vector<float> fbinningY(binningY.begin(),binningY.end());
+  float* arrY = &fbinningY[0];
+
+  me.numerator   = ibooker.book2D(histname+"_numerator",   histtitle+" (numerator)",   nbinsX, arrX, nbinsY, arrY);
+  me.denominator = ibooker.book2D(histname+"_denominator", histtitle+" (denominator)", nbinsX, arrX, nbinsY, arrY);
+}
+
+void BPHMonitor::bookHistograms(DQMStore::IBooker     & ibooker,
+				 edm::Run const        & iRun,
+				 edm::EventSetup const & iSetup) 
+{  
+  
+  std::string histname, histtitle;
+
+  std::string currentFolder = folderName_ ;
+  ibooker.setCurrentFolder(currentFolder.c_str());
+  
+  histname = "muPt"; histtitle = "mu_P_{t}";
+  bookME(ibooker,muPt_,histname,histtitle, pt_binning_.nbins, pt_binning_.xmin, pt_binning_.xmax);
+  setMETitle(muPt_,"Mu_Pt[GeV]","events/1GeV");
+
+  histname = "muPhi"; histtitle = "muPhi";
+  bookME(ibooker,muPhi_,histname,histtitle, phi_binning_.nbins, phi_binning_.xmin, phi_binning_.xmax);
+  setMETitle(muPhi_," mu_#phi","events / 0.1 rad");
+
+  histname = "muEta"; histtitle = "mu_Eta";
+  bookME(ibooker,muEta_,histname,histtitle, eta_binning_.nbins,eta_binning_.xmin, eta_binning_.xmax);
+  setMETitle(muEta_," mu_#eta","events/ ");
+
+  histname = "mu_d0"; histtitle = "mu_d0";
+  bookME(ibooker,mud0_,histname,histtitle, d0_binning_.nbins,d0_binning_.xmin, d0_binning_.xmax);
+  setMETitle(mud0_," mu_d0","events/bin ");
+
+  histname = "mu_z0"; histtitle = "mu_z0";
+  bookME(ibooker,muz0_,histname,histtitle, z0_binning_.nbins,z0_binning_.xmin, z0_binning_.xmax);
+  setMETitle(muz0_," mu_z0","events/bin ");
+
+  // Initialize the GenericTriggerEventFlag
+  if ( num_genTriggerEventFlag_ && num_genTriggerEventFlag_->on() ) num_genTriggerEventFlag_->initRun( iRun, iSetup );
+  if ( den_genTriggerEventFlag_ && den_genTriggerEventFlag_->on() ) den_genTriggerEventFlag_->initRun( iRun, iSetup );
+
+}
+
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
+#include "Geometry/Records/interface/TrackerTopologyRcd.h"
+void BPHMonitor::analyze(edm::Event const& iEvent, edm::EventSetup const& iSetup)  {
+
+//  edm::Handle<reco::BeamSpot> const& beamSpot;
+  // Filter out events if Trigger Filtering is requested
+  if (den_genTriggerEventFlag_->on() && ! den_genTriggerEventFlag_->accept( iEvent, iSetup) ) return;
+
+  
+  edm::Handle<reco::BeamSpot> beamSpot;
+  iEvent.getByToken( bsToken_,  beamSpot);
+  edm::Handle<reco::MuonCollection> muoHandle;
+  iEvent.getByToken( muoToken_, muoHandle );
+
+  if ( int(muoHandle->size()) < nmuons_ ) return;
+  std::vector<reco::Muon> muons;
```

Comment:
why is this copy of of the data product needed? Looks like it is just accessed as a const below?